new-log-viewer/src/contexts/NotificationContextProvider.tsx

```diff
@@ -0,0 +1,107 @@
+import React, {
+    createContext,
+    useCallback,
+    useRef,
+    useState,
+} from "react";
+
+import {
+    Nullable,
+    WithId,
+} from "../typings/common";
+import {LOG_LEVEL} from "../typings/logs";
+
+
+interface PopupMessage {
+    level: LOG_LEVEL,
+    message: string,
+    timeoutMillis: Nullable<number>,
+    title: string,
+}
+
+interface NotificationContextType {
+    popupMessages: WithId<PopupMessage>[],
+
+    handlePopupMessageClose: (messageId: number) => void;
+    postPopup: (message: PopupMessage) => void,
+}
+
+const NotificationContext = createContext<NotificationContextType>({} as NotificationContextType);
+
+/**
+ * Default values of the Notification context value object.
+ */
+const NOTIFICATION_DEFAULT: Readonly<NotificationContextType> = Object.freeze({
+    popupMessages: [],
+
+    handlePopupMessageClose: () => {},
+    postPopup: () => {},
+});
+
+/**
+ * The default duration in milliseconds after which an automatic dismissal will occur.
+ */
+const DEFAULT_AUTO_DISMISS_TIMEOUT_MILLIS = 10_000;
+
+/**
+ * A value that indicates that a pop-up message should not be automatically dismissed.
+ */
+const DO_NOT_TIMEOUT_VALUE = null;
+
+interface NotificationContextProviderProps {
+    children: React.ReactNode,
+}
+
+/**
+ * Provides notification management for the application. This provider must be at the outermost
+ * layer of subscriber / publisher components to ensure they can receive / publish notifications.
+ *
+ * @param props
+ * @param props.children
+ * @return
+ */
+const NotificationContextProvider = ({children}: NotificationContextProviderProps) => {
+    const [popupMessages, setPopupMessages] = useState<WithId<PopupMessage>[]>(
+        NOTIFICATION_DEFAULT.popupMessages
+    );
+    const nextPopUpMessageIdRef = useRef<number>(0);
+
+    const postPopup = useCallback((message:PopupMessage) => {
+        const newMessage = {
+            id: nextPopUpMessageIdRef.current,
+            ...message,
+        };
+
+        nextPopUpMessageIdRef.current++;
+
+        setPopupMessages((v) => ([
+            ...v,
+            newMessage,
```

Comment:
I do see how we don't want any existing message to move after any new ones are added. However, I think there's one overflowing issue we need to deal with when there're too many pop-up messages (if the new messages appear on the top when they're more than what the viewport can display, they could be hidden). Let's see if we can fix the behaviour in this PR; if not, I think it's fine to leave the new message at the bottom and file the issue which is to be fixed later.

Reply:
i'm thinking about adding a transparent container (and/or semi-transparent when being hovered over?) to hold the message boxes. By making the container scrollable and to have the same height of the viewport, it should make all messages scrollable. What do you think?

Reply:
![image](https://github.com/user-attachments/assets/6ff3d671-f5e7-4fd0-93d2-9c43636ed6b3)

done. Now we make the notifications scrollable if there're more than what the viewport can display in height.

I did one CSS change `scrollbar-width: none;` to hide the scroll bar visually while keeping the container scrollable. If we want to show the scrollbar we can keep that.

As for the new message location, I discussed with Kirk offline and we believe placing the new message at the bottom can avoid us having to implement the "auto-scroll-to-top" behaviour (technically it's possible but we might not want to simply always scroll to the top every time there're new messages added / deleted). Feel free to object

Reply:
Note that the new messages are placed at the front of the array though visually they appear at the bottom because now we have `column-reverse` order set on the Stack.